### ISSUE TITLE
Error while training fcn32s

### ISSUE THREAD

**Issue by pyrrhicpk** [vlfeat/matconvnet-fcn#5]
Hi,

I am trying to train the fcn32s using my own data for segmentation purposes. I load the "imagenet-vgg-verydeep-16.mat" model and run the function "fcnInitializeModel" to create the fcn32 model.
My training set is a 224X224X3XN matrix (rgb images) and labels set is 224X224X1XN matrix (pixel-wise labels). However, I encounter the following error whenever I run the command: info = cnn_train_dag(net, imdb, @getBatch,opts.train,'val',find(imdb.images.set == 3));

The error is:

Subscript indices must either be real positive integers or logicals.
Error in dagnn.DagNN/eval (line 74)
[obj.vars(v).value] = deal(inputs{2:2:end}) ;
Error in cnn_train_dag>process_epoch (line 179)
      net.eval(inputs, opts.derOutputs) ;
Error in cnn_train_dag (line 84)
    stats.train(epoch) = process_epoch(net, state, opts, 'train') ;

Does anybody know how to fix this error?

Thanks,
Best regards

**Comment by vedaldi** [comment]
Hi,

it is probably a typo on the name of the input variables that the code is passing to the DAG. If you `dbstop on error` and check the index `v`, it may be NaN, indicating that the variable name was not found. We are improving the error messages as some are pretty confusing.

Best.,

Andrea

**Comment by pyrrhicpk** [comment]
Hi Andrea,

Thanks for the information. Yes, the variable 'v' gets a NaN value. After correcting the input variable name, the code now prompts a different error:

Error using  - 
Matrix dimensions must agree.
Error in cnn_train_dag>accumulate_gradients (line 234)
  state.momentum{i} = opts.momentum * state.momentum{i} ...
Error in cnn_train_dag>process_epoch (line 195)
    state = accumulate_gradients(state, net, opts, batchSize, mmap) ;
Error in cnn_train_dag (line 84)
    stats.train(epoch) = process_epoch(net, state, opts, 'train') ;

Could it be due wrong dimensions of my data or labels? Currently, my imdb.images.data is 224x224x3xN, imdb.images.labels is 224x224x1xN, and imdb.images.set is 1xN matrices. I hope I am using the correct data format.

Thanks,
Best regards

**Comment by vedaldi** [comment]
Hi,

I believe that the current implementation requires at least two labels. I.e. for a binary problem, you should still have a 224x224x2xN mask. This could very well be the reason.

Best wishes,

Andrea

**Comment by pyrrhicpk** [comment]
Hi,

I have re-built my imdb with respect to 7 classes so that imdb.images.data is a 4-D single 224x224x3xN, imdb.images.labels is a 4-D single 224x224x7xN, set is a 1xN and data_mean is a 224x224x3 matrix. However, the code still prompts the same matrix dimensions error as before:

Error using  - 
Matrix dimensions must agree.
Error in cnn_train_dag>accumulate_gradients (line 234)
  state.momentum{i} = opts.momentum * state.momentum{i} ...
Error in cnn_train_dag>process_epoch (line 195)
    state = accumulate_gradients(state, net, opts, batchSize, mmap) ;
Error in cnn_train_dag (line 84)
    stats.train(epoch) = process_epoch(net, state, opts, 'train') ;

Any other workaround?

Thanks,
Best regards

**Comment by ZhichenZhao** [comment]
Hi, @pyrrhicpk ,
the size of label before training I think should be uint8, like 224x224, and if you have 7 classes, elements in this label ranges from 0 to 6. Before this assertion it should change to be 224x224x1xN where N is the number of this batch. Hope to help.
But my net can not converge, and I don't know why. Maybe we can discuss later.

Best regards.
Zhichen

**Comment by pyrrhicpk** [comment]
Hi,

Thanks Zhichen for pointing out that. Now, with imdb.images.labels to be a 224x224x1xN matrix, I am stuck with the following error:

Index exceeds matrix dimensions.
Error in vl_nnloss (line 194)
      t = Xmax + log(sum(ex,3)) - X(ci) ;

Here, both X and ci have the same dimensions (i.e., 224x224x1x15), where 15 is my batch size. I noticed that X contains all zeros in all 15 channels, whereas ci contains values. Any idea what's going wrong?

Thanks

**Comment by pyrrhicpk** [comment]
Hi,

Thanks Zhichen for pointing out that. Now, with 224x224x1xN format for the labels, I am stuck with another error:

Error using gpuArray/subsref
Index exceeds matrix dimension.
Error in vl_nnloss (line 194)
       t = Xmax + log(sum(ex,3)) - X(ci) ;

Here, Both X and ci have 224x224x1x15 dimensions (where 15 is my batch size). I noticed that X contains all zeros whereas ci contains values. Any idea what's going wrong?

Thanks

**Comment by ZhichenZhao** [comment]
Hi @pyrrhicpk ,
I don't know what the error is. Have you run the code on PASCAL VOC firstly? I think you could set your data and label as the same format as in VOC to make sure it could work on your own data. To this error, I guess it happened out of vl_nnloss, you may use "dbstop if error" or any else to check.
Good luck

**Comment by brisker** [comment]
Hi   @vedaldi   When I am running the fcnTrain.m the speed is very very slow and not normal, what problems do you think I have?
the code running like this:
CUDADevice with properties:

```
              Name: 'GeForce GTX TITAN X'
             Index: 1
 ComputeCapability: '5.2'
    SupportsDouble: 1
     DriverVersion: 7.5000
    ToolkitVersion: 6.5000
MaxThreadsPerBlock: 1024
  MaxShmemPerBlock: 49152
MaxThreadBlockSize: [1024 1024 64]
       MaxGridSize: [2.1475e+09 65535 65535]
         SIMDWidth: 32
       TotalMemory: 1.2885e+10
   AvailableMemory: 1.2609e+10
```

   MultiprocessorCount: 24
          ClockRateKHz: 1076000
           ComputeMode: 'Default'
  GPUOverlapsTransfers: 1
KernelExecutionTimeout: 1
      CanMapHostMemory: 1
       DeviceSupported: 1
        DeviceSelected: 1
train: epoch 01: 1/565: 0.9 Hz accuracy: 0.677 0.048 0.032 objective: 3.044
train: epoch 01: 2/565: 1.0 Hz accuracy: 0.691 0.048 0.033 objective: 3.035
train: epoch 01: 3/565: 1.0 Hz accuracy: 0.698 0.048 0.033 objective: 2.994

**Comment by physics-programmer** [comment]
@pyrrhicpk this error is due to data and label class, which must be set to uint8 : 
images.data(:,:,:,j) = uint8(image);
images.label(:,:,1,j) = uint8(mask);  
So I run a modified fcnTrain on 1 class (person), network seems to train, but have an accuracy of 1 during training : 


train: epoch 01:   1/  4: 0.4 (0.4) Hz accuracy: 1.000 4.761905e-02: 0.048 objective: 0.693
train: epoch 01:   2/  4: 0.4 (0.4) Hz accuracy: 1.000 4.761905e-02: 0.048 objective: 0.685
train: epoch 01:   3/  4: 0.4 (0.4) Hz accuracy: 1.000 4.761905e-02: 0.048 objective: 0.672
train: epoch 01:   4/  4: 0.4 (0.4) Hz accuracy: 1.000 4.761905e-02: 0.048 objective: 0.654

[net-train.pdf](https://github.com/vlfeat/matconvnet-fcn/files/806342/net-train.pdf)

**Comment by BigBugX** [comment]
Hi all,

I've met a similar problem, and do not know what to do with it. The error info is as follows:

train: epoch 01:   1/ 11:Error using dagnn.DagNN/eval (line 80)
No variable of name 'input' could be found in the DAG.

Error in cnn_train_dag>process_epoch (line 213)
      net.eval(inputs, opts.derOutputs) ;

Error in cnn_train_dag (line 83)
    [stats.train(epoch),prof] = process_epoch(net, state, opts, 'train') ;

Error in cnn_four (line 61)
[net, info] = trainFn(net, imdb, getBatchFn(opts, net.meta), ...

I noticed that the variable "inputs" is actually a cell containing a batch of images and their labels (the detail is like: inputs = {'inputs', IMAGES, 'labels', LABELS}, while IMAGES/LABELS are images and corresponding labels respectively), but I just don't understand why "getVarIndex" is try to get the 'inputs' in cell inputs? this string ('inputs') could always be NaN and the code always gets wrong here!  Could any body give me some advice?

Thanks and Regards
